src/ast/semantic_analyser.cpp

```diff
@@ -1337,6 +1330,21 @@ void SemanticAnalyser::visit(Binop &binop)
               << "(cast to unsigned to silence warning)";
         }
       }
+
+      auto pos_param = dynamic_cast<PositionalParameter *>(left);
```

Comment:
`right` can also be positional params (e.g., `str(1 + $1)`). Looks good to me other than that.

Reply:
I realized that we cannot write something like `str($1 + 2*4)`. I also found that we can bypass the check of addition length by using multiple add (e.g., `str($1 + 3 + 3)`.) But I guess it is probably hard to cope with these problems, so I think the current behavior is OK.

Reply:
what is the usecase for this?

I don't think we should allow arbitrary complex expressions in this. A single literal integer as offset should be enough. If we ever land the option to reduce expressions (e.g. `1+2*3` into `7`) we can always add it back in.

Reply:
I agree, a literal integer offset should be enough.

Reply:
I also agree, a single constant offset should be sufficient. I tried to make the semantic checks more robust though, so currently only `str($1)`, `str($1 + CONST)`, or `str(CONST + $1)` are supported, everything else should cause a semantic error.